Add documentation for `global_control::leave_policy` (#2045)
oneapi-src/oneTBB: doc/main/reference/parallel_phase_for_task_arena.rst

Reviewer: ```suggestion
    To enable this feature, set ``TBB_PREVIEW_PARALLEL_PHASE`` macro to 1. When available and enabled, the feature-test macro ``TBB_HAS_PARALLEL_PHASE`` is defined.
```

Description
***********

> By default, oneTBB uses a *delayed leave* heuristic: after completing work in an arena,

Reviewer 2: ```suggestion
By default, oneTBB uses a *delayed thread leave* heuristic: after completing work in an arena,
```

worker threads remain for an implementation-defined duration, anticipating that new parallel
work will arrive soon. This benefits most workloads by reducing the latency of starting
subsequent parallel computations. However, the heuristic can hurt performance in two scenarios:

* **Unpredictable work patterns.** When parallel tasks are submitted at irregular intervals or
  with long gaps, retaining idle workers wastes CPU resources.
* **Composability with other parallel runtimes.** When oneTBB is interleaved with another
  runtime (for example, OpenMP), retained workers can cause oversubscription and reduce

>   overall throughput.

Reviewer 2: ```suggestion
subsequent parallel computations. However, this behavior can be undesirable, especially if

* parallel tasks are submitted at irregular intervals or with long gaps, and idle threads waste CPU resources;
* oneTBB use is interleaved with another threading, and idle threads cause CPU oversubscription.
```

This feature gives users explicit control over worker thread retention. A *leave policy*

> determines how quickly workers leave an arena when no work is available. Additionally, the

Reviewer 2: ```suggestion
For explicit control over worker thread retention, a *leave policy* determines
how fast worker threads leave an arena when no work is available. Additionally, the
```

*parallel phase* API lets users bracket regions of parallel work so the scheduler can retain

> workers more aggressively during those regions and release them promptly afterward.

Reviewer 2: ```suggestion
*parallel phase* API lets users bracket regions of recurrent parallel work so the scheduler can
retain threads more aggressively during those regions and release them promptly afterward.
```

This feature extends the :onetbb-spec:`tbb::task_arena specification <task_scheduler/task_arena/task_arena_cls>`
with the following API:

* Adds the ``leave_policy`` enumeration class to ``task_arena``.
* Adds ``leave_policy`` as the last parameter in ``task_arena`` constructors and ``task_arena::initialize`` methods.
  This allows you to inform the scheduler about the preferred policy for worker threads
  when they are about to leave ``task_arena`` due to a lack of available work.
* Adds new ``start_parallel_phase`` and ``end_parallel_phase`` interfaces to the ``task_arena`` class
  and the ``this_task_arena`` namespace. These interfaces work as hints to the scheduler to mark the start and end
  of parallel work submission into the arena, enabling different worker thread retention policies.
* Adds the Resource Acquisition is Initialization (RAII) class ``scoped_parallel_phase`` to ``task_arena``.

> * Adds the ``leave_policy`` parameter to the ``global_control`` class, providing application-wide
>   control over the default worker thread leave behavior for arenas initialized with
>   ``leave_policy::automatic``.

Reviewer 2: ```suggestion
* Adds the ``leave_policy`` parameter to the ``global_control`` class, providing application-wide
  control over the default worker thread leave behavior for arenas initialized implicitly or with
  ``leave_policy::automatic``.
```

  and the ``this_task_arena`` namespace. These interfaces work as hints to the scheduler to mark the start and end
  of parallel work submission into the arena, enabling different worker thread retention policies.
* Adds the Resource Acquisition is Initialization (RAII) class ``scoped_parallel_phase`` to ``task_arena``.
* Adds the ``leave_policy`` parameter to the ``global_control`` class, providing application-wide
  control over the default worker thread leave behavior for arenas initialized with
  ``leave_policy::automatic``.

More details about motivation, semantics and conditions for becoming fully supported functionality can be found in the corresponding
`Request For Comments document for parallel_phase <https://github.com/uxlfoundation/oneTBB/tree/master/rfcs/experimental/parallel_phase_for_task_arena>`_

> and the `Request For Comments document for global_control::leave_policy <https://github.com/uxlfoundation/oneTBB/tree/master/rfcs/experimental/global_control_fast_leave>`_.

Reviewer 2: I'd suggest to take all necessary details to the documentation and remove these links.

Author: Added new motivation and semantics paragraphs and removed the RFC links.

Indicates the end of the parallel phase in the current ``task_arena``.
If ``with_fast_leave`` is ``true``, worker threads leave policy is temporarily set to ``fast``.

Global Control Integration
--------------------------

.. cpp:enum:: global_control::leave_policy

> **Selection rule**: logical disjunction

Reviewer 2: I am still uneasy about describing the rule as logical disjunction for a parameter which values are not Boolean.

Instead, I would suggest to either move up the more detailed description from below, or use "see below" here.

Author: Adding "see below" makes sense. Updated.

When the ``leave_policy`` parameter is active on a ``global_control`` object with
the value ``task_arena::leave_policy::fast``, initializing an arena with
``task_arena::leave_policy::automatic`` behaves as if the arena is initialized with

> ``task_arena::leave_policy::fast``. Arenas that were already initialized are not affected 

Reviewer 3: ```suggestion
``task_arena::leave_policy::fast``. Arenas that were already initialized (including implicit arenas) are not affected 
```

by changes to the ``leave_policy`` parameter on a ``global_control`` object.

When multiple ``global_control`` objects exist for the ``leave_policy`` parameter,
their values are combined using logical disjunction: if any active object has the

> value ``task_arena::leave_policy::fast``, fast leave is enabled globally.

Reviewer 2: ```suggestion
When multiple ``global_control`` objects exist for the ``leave_policy`` parameter,
their values are combined as follows: the active parameter value equals to
``task_arena::leave_policy::fast`` if any alive ``global_control`` object sets that value,
otherwise it equals to ``task_arena::leave_policy::automatic``.
```

> The following table summarizes the interaction between the per-arena and global leave policies:

Reviewer 2: ```suggestion
The following table summarizes the interaction between the per-arena and global leave policies
when an arena is created:
```

The following table summarizes the interaction between the per-arena and global leave policies
when an arena is created:

.. table::

    +------------------------+-------------------------+------------------------+
    | Arena ``leave_policy`` | Global ``leave_policy`` | Initial State          |
    +========================+=========================+========================+
    | ``fast``               | any                     | Fast leave             |
    +------------------------+-------------------------+------------------------+
    | ``automatic``          | ``fast``                | Fast leave             |
    +------------------------+-------------------------+------------------------+
    | ``automatic``          | ``automatic`` (default) | System-specific policy |
    +------------------------+-------------------------+------------------------+
.. note::
   The ``global_control::leave_policy`` parameter provides application-wide control,
   while ``task_arena::leave_policy`` and ``parallel_phase`` provide per-arena control.
   After initialization, the parallel phase API can independently modify the arena's

>    leave behavior at runtime, regardless of the initial state set by the global control.

Reviewer 2: ```suggestion
   After arena initialization, the parallel phase API can modify the thread leave behavior
   for the arena at runtime, regardless of the initial state set by the global control.
```
